### ISSUE TITLE
Refresh does not update stack outputs

### ISSUE THREAD

**Issue by melbourne2991** [pulumi/pulumi#2710]
It seems as though some properties are not updated when they are changed in AWS, even after doing a refresh. When you stop an ec2 instance and start it again, AWS may give it a difference public ip - Pulumi appears to still display the original public ip.

1. Spin up an ec2 instance with Pulumi, make note of output publicIp/publicDns
2. Stop the instance via AWS console,
3. Start the instance again via the AWS console
4. Run `pulumi refresh`
5. Run `pulumi stack output` and compare publicIp in aws console with the one displayed in Pulumi's output

**Comment by lukehoban** [comment]
> Run pulumi refresh

When I do this, I get this (as expected):

```
   ~ aws:ec2/instance:Instance: (update)
        [id=i-049724aab220cb841]
        [urn=urn:pulumi:dev::pulumi2710::aws:ec2/instance:Instance::instance]
        --outputs:--
        ami                      : "ami-0ff8a91507f77f867"
        arn                      : "arn:aws:ec2:us-east-1:153052954103:instance/i-049724aab220cb841"
        associatePublicIpAddress : true
        availabilityZone         : "us-east-1c"
        cpuCoreCount             : 1
        cpuThreadsPerCore        : 2
        creditSpecification      : {
            cpuCredits: "unlimited"
        }
        disableApiTermination    : false
        ebsOptimized             : false
        getPasswordData          : false
        id                       : "i-049724aab220cb841"
        instanceState            : "running"
        instanceType             : "t3.nano"
        monitoring               : false
        primaryNetworkInterfaceId: "eni-0bda07c18d7ce3045"
        privateDns               : "ip-172-31-22-132.ec2.internal"
        privateIp                : "172.31.22.132"
      ~ publicDns                : "ec2-54-91-29-152.compute-1.amazonaws.com" => "ec2-54-174-106-19.compute-1.amazonaws.com"
      ~ publicIp                 : "54.91.29.152" => "54.174.106.19"
        rootBlockDevice          : {
            deleteOnTermination: true
            iops               : 100
            volumeId           : "vol-058974c4a07b71edf"
            volumeSize         : 8
            volumeType         : "gp2"
        }
        securityGroups           : [
            [0]: "default"
        ]
        sourceDestCheck          : true
        subnetId                 : "subnet-e1a00ca9"
        tenancy                  : "default"
        vpcSecurityGroupIds      : [
            [0]: "sg-a440bbda"
        ]
```

> Run pulumi stack output and compare publicIp in aws console with the one displayed in Pulumi's output

Indeed, when I run this, I get the wrong thing:

```
$ pulumi stack output
Current stack outputs (1):
    OUTPUT      VALUE
    bucketName  54.91.29.152
```

But if I then do another `pulumi update`, it makes no changes by does produce the newly correct outputs:

```
$ pulumi up
...
Outputs:
  ~ bucketName: "54.91.29.152" => "54.174.106.19"
```

**Comment by lukehoban** [comment]
I believe the issue here is that `pulumi refresh` does not recompute the outputs.  In fact, `pulumi refresh` does not run the program at all, so it cannot re-flow the implications of changes through to outputs.

Running `pulumi update` again will fix this (by re-running the deployment and ensuring that outputs are re-populated).  But this is surprising behaviour in general, and we may want to do something to encourage re-running `pulumi up` to flow changes through into outputs if that is the only way.

**Comment by iamim** [comment]
Is there any way of reconciling the state that is currently in the cloud with the state I get inside of pulumi script?

This would be very useful to address some scenarios that are hard to achieve otherwise. For example, see this discussion: https://github.com/Azure/bicep/discussions/10671
TL;DR: it's not possible (afaik) to deploy a new revision of an Azure Container App without explicitly listing all of the traffic balancer weight with the default being set 100% to the latest deployment. Now, it is possible to achieve with Pulumi by keeping the `traffic` property in the output.

Additionally, I would love to enable my team (including those in less technical roles) to perform green-blue like deployments without having to modify the pulumi program.

**Comment by iamim** [comment]
I found a workaround to achieve that by keeping the app name in the output querying the current state of the ACA in question. It does not feel great, but it's something.

The only thing I'm not certain of is how Pulumi would behave when it's running for the first time, i.e. what would happen if `azn.app.ContainerApp.get(...)` throws because the application ID does not exist yet.

Also, after facing this issue I realized that it's not 100% clear to me what is the difference between `pulumi up` `pulumi up -r` and `pulumi refresh`. Perhaps it could be documented better or am I missing something?

The code for reference
```typescript
// TODO: how to manage the first deployment?
const queriedApp = azn.app.ContainerApp.get("queriedApp", self.outputs.apply(o => o.app?.id));
export const queriedTraffic = queriedApp.configuration.apply(c => c?.ingress?.traffic);
export const queriedAppName = queriedApp.name;

function getTraffic(newRevisionSuffix: string): pulumi.Output<azn.types.input.app.TrafficWeightArgs[]> {
    return pulumi.all([queriedAppName, queriedTraffic]).apply(([appName, t]) => {
        if (t === undefined || appName === undefined) {
            if (appName === undefined) {
                return pulumi.output([]);
            }

            const newRevision = `${appName}--${newRevisionSuffix}`;
            const first: azn.types.input.app.TrafficWeightArgs = {
                label: "latest",
                revisionName: newRevision,
                weight: 100,
            };

            return pulumi.output([first]);
        }

        const newRevision = `${appName}--${newRevisionSuffix}`;
        const filtered = t.filter(r => r.revisionName && r.revisionName !== newRevision && r.weight !== 0);
        
        // delete latest label
        filtered.forEach(r => {
            if (r.label === "latest") {
                r.label = undefined;
            }
        });

        return pulumi.output([
            ...filtered,
            {
                label: "latest",
                revisionName: newRevision,
                weight: filtered.length === 0 ? 100 : 0,
            }
        ]);
    });
}
```

**Comment by Frassle** [comment]
> The only thing I'm not certain of is how Pulumi would behave when it's running for the first time, i.e. what would happen if azn.app.ContainerApp.get(...) throws because the application ID does not exist yet.

It won't throw it will just stop the deployment engine and ignore the rest of the program. There's separate issues about changing that (https://github.com/pulumi/pulumi/issues/3364).

**Comment by lunaris** [comment]
I _think_ we can consider this fixed now that we have `--run-program` for `refresh`, which runs the program (and thus recomputes stack outputs), but feel free to reopen if I've missed something 🙏
